stacks-signer/src/signer.rs

```diff
@@ -1046,23 +1038,34 @@ impl Signer {
     /// Update the DKG for the provided signer info, triggering it if required
     pub fn update_dkg(&mut self, stacks_client: &StacksClient) -> Result<(), ClientError> {
         let reward_cycle = self.reward_cycle;
-        let aggregate_public_key = stacks_client.get_aggregate_public_key(reward_cycle)?;
-        self.coordinator
-            .set_aggregate_public_key(aggregate_public_key);
+        let new_aggregate_public_key = stacks_client.get_aggregate_public_key(reward_cycle)?;
+        let old_aggregate_public_key = self.coordinator.get_aggregate_public_key();
+        if new_aggregate_public_key.is_some()
+            && old_aggregate_public_key != new_aggregate_public_key
+        {
+            debug!(
+                "Signer #{}: Received a new aggregate public key ({new_aggregate_public_key:?}) for reward cycle {reward_cycle}. Overwriting its internal aggregate key ({old_aggregate_public_key:?})",
+                self.signer_id
+            );
+            self.coordinator
+                .set_aggregate_public_key(new_aggregate_public_key);
+        }
         let coordinator_id = stacks_client
             .calculate_coordinator(&self.signing_round.public_keys)
             .0;
-        if aggregate_public_key.is_none()
+        // TOOD: should check if we have an aggregate key before we trigger another round. may have a delay between us calculating a key, broadcasting to the contract, and it being confirmed by the miner.
+        if new_aggregate_public_key.is_none()
             && self.signer_id == coordinator_id
             && self.coordinator.state == CoordinatorState::Idle
+        // Have I already voted and have a pending transaction? Check stackerdb for the same round number and reward cycle
         {
             info!("Signer #{} is the current coordinator for {reward_cycle}. Triggering a DKG round...", self.signer_id);
             if self.commands.back() != Some(&Command::Dkg) {
                 self.commands.push_back(Command::Dkg);
             }
```

Comment:
Is this a sufficient check for what we need here? Couldn't other commands have come in during `run_one_pass` before calling `refresh_signers` -> `update_dkg`?

Reply:
Yes they could have. Have actually changed this to check the front. If we havent run dkg and its the next thing we need to run, don't bother appending. Otherwise run it. If other commands come in (they are from an external thread) or are just sign commands and we cannot sign if we haven't done dkg yet first.